executor/simple.go

```diff
@@ -1048,6 +1050,123 @@ func (e *SimpleExec) executeGrantRole(s *ast.GrantRoleStmt) error {
 	return nil
 }
 
+// Should cover same internal mysql.* tables as DROP USER, so this function is very similar
+func (e *SimpleExec) executeRenameUser(s *ast.RenameUserStmt) error {
+
+	failedUsers := make([]string, 0, len(s.UserToUsers))
+	sysSession, err := e.getSysSession()
+	defer e.releaseSysSession(sysSession)
+	if err != nil {
+		return err
+	}
+	sqlExecutor := sysSession.(sqlexec.SQLExecutor)
+
+	if _, err := sqlExecutor.ExecuteInternal(context.TODO(), "begin"); err != nil {
+		return err
+	}
+
+	for _, userToUser := range s.UserToUsers {
+		oldUser, newUser := userToUser.OldUser, userToUser.NewUser
+		exists, err := userExistsInternal(sqlExecutor, oldUser.Username, oldUser.Hostname)
+		if err != nil {
+			return err
+		}
+		if !exists {
+			failedUsers = append(failedUsers, oldUser.String()+" TO "+newUser.String()+" old did not exist")
+			break
+		}
+
+		exists, err = userExistsInternal(sqlExecutor, newUser.Username, newUser.Hostname)
+		if err != nil {
+			return err
+		}
+		if exists {
+			// MySQL reports the old user, even when the issue is the new user.
+			failedUsers = append(failedUsers, oldUser.String()+" TO "+newUser.String()+" new did exist")
+			break
+		}
+
+		if err = renameUserHostInSystemTable(sqlExecutor, mysql.UserTable, "User", "Host", userToUser); err != nil {
```

Comment:
There are 9 system tables to modify. Maybe you can put the params in an array and iterate the array to simplify the code?

Reply:
I agree, and that is why I also [added a comment](https://github.com/pingcap/tidb/pull/24413/files#diff-96117c390a5a6cfe0ea63921e2252de41e74ea40e00da756907c7b55876e0197R1141) about it.

Do you want me to add the refactoring of DROP USER and the new code in this PR or in a separate PR?

Reply:
I prefer a separate PR.